aws_lambda_powertools/utilities/idempotency/persistence/redis.py

```diff
@@ -0,0 +1,440 @@
+
+            redis_response = self.client.set(name=item["name"], value=encoded_item, ex=ttl, nx=True)
+
+            # redis_response:True -> Redis set succeed, idempotency key does not exist before
+            # return to idempotency and proceed to handler execution phase. Most cases should return here
+            if redis_response:
+                return
+
+            # redis_response:None -> Existing record on Redis, continue to checking phase
+            # The idempotency key exist:
+            #    - previous invocation with the same key and not expired(active idempotency)
+            #    - previous invocation timed out (Orphan Record)
+            #    - previous invocation record expired but not deleted by Redis (Orphan Record)
+
+            idempotency_record = self._get_record(item["name"])
+
+            # status is completed and expiry_attr timestamp still larger than current timestamp
+            # found a valid completed record
+            if idempotency_record.status == STATUS_CONSTANTS["COMPLETED"] and not idempotency_record.is_expired:
+                raise IdempotencyItemAlreadyExistsError
+
+            # in_progress_expiry_attr exist means status is in_progress, and still larger than current timestamp,
+            # found a vaild in_progress record
+            if (
+                idempotency_record.status == STATUS_CONSTANTS["INPROGRESS"]
+                and idempotency_record.in_progress_expiry_timestamp
+                and idempotency_record.in_progress_expiry_timestamp > int(now.timestamp() * 1000)
+            ):
+                raise IdempotencyItemAlreadyExistsError
+
+            # If the code reaches here means we found an Orphan record.
+            raise IdempotencyOrphanRecordError
+
+        except IdempotencyOrphanRecordError:
+            # deal with orphan record here
+            # aquire a lock for default 10 seconds
+            with self._acquire_lock(name=item["name"]):
+                self.client.set(name=item["name"], value=encoded_item, ex=ttl)
+
+            # lock was not removed here intentionally. Prevent another orphan operation in race condition.
+        except (redis.exceptions.RedisError, redis.exceptions.RedisClusterException) as e:
+            raise e
+        except Exception as e:
+            logger.debug(f"encountered non-redis exception:{e}")
+            raise e
+
+    @contextmanager
+    def _acquire_lock(self, name: str):
+        """
+        aquire a lock for default 10 seconds
+        """
+        try:
+            acquired = self.client.set(name=f"{name}:lock", value="True", ex=self._orphan_lock_timeout, nx=True)
+            logger.debug("acquiring lock to overwrite orphan record")
+            if acquired:
+                logger.debug("lock acquired")
+                yield
+            else:
+                # lock failed to aquire, means encountered a race condition. Just return
```

Comment:
```suggestion
                # If the lock acquisition fails, it suggests a race condition has occurred. In this case, instead of
                # proceeding, we log the event and raise an error to indicate that the current operation should be
                # retried after the lock is released by the process that currently holds it.
```